fourier/rfft.go

```diff
@@ -0,0 +1,1134 @@
+	cc3 := newThreeArray(ido, ip, l1, cc)
+	c13 := newThreeArray(ido, l1, ip, c1)
+	ch3 := newThreeArray(ido, l1, ip, ch)
+	c2m := newTwoArray(idl1, ip, c2)
+	ch2m := newTwoArray(idl1, ip, ch2)
+
+	arg := 2 * math.Pi / float64(ip)
+	dcp := math.Cos(arg)
+	dsp := math.Sin(arg)
+	ipph := (ip + 1) / 2
+	ipp2 := ip + 2
+	idp2 := ido + 2
+	nbd := (ido - 1) / 2
+
+	if ido == 1 {
+		for ik := 1; ik <= idl1; ik++ {
+			c2m.set(ik, 1, ch2m.at(ik, 1))
+		}
+	} else {
+		for ik := 1; ik <= idl1; ik++ {
+			ch2m.set(ik, 1, c2m.at(ik, 1))
+		}
+		for j := 2; j <= ip; j++ {
+			for k := 1; k <= l1; k++ {
+				ch3.set(1, k, j, c13.at(1, k, j))
+			}
+		}
+
+		is := -ido
+		if nbd > l1 {
+			for j := 2; j <= ip; j++ {
+				is += ido
+				for k := 1; k <= l1; k++ {
+					idij := is
+					for i := 3; i <= ido; i += 2 {
+						idij += 2
+						ch3.set(i-1, k, j, wa.at(idij-1)*c13.at(i-1, k, j)+wa.at(idij)*c13.at(i, k, j))
+						ch3.set(i, k, j, wa.at(idij-1)*c13.at(i, k, j)-wa.at(idij)*c13.at(i-1, k, j))
+					}
+				}
+			}
+		} else {
+			for j := 2; j <= ip; j++ {
+				is += ido
+				idij := is
+				for i := 3; i <= ido; i += 2 {
+					idij += 2
+					for k := 1; k <= l1; k++ {
+						ch3.set(i-1, k, j, wa.at(idij-1)*c13.at(i-1, k, j)+wa.at(idij)*c13.at(i, k, j))
+						ch3.set(i, k, j, wa.at(idij-1)*c13.at(i, k, j)-wa.at(idij)*c13.at(i-1, k, j))
+					}
+				}
+			}
+		}
+		if nbd < l1 {
+			for j := 2; j <= ipph; j++ {
+				jc := ipp2 - j
+				for i := 3; i <= ido; i += 2 {
+					for k := 1; k <= l1; k++ {
+						c13.set(i-1, k, j, ch3.at(i-1, k, j)+ch3.at(i-1, k, jc))
+						c13.set(i-1, k, jc, ch3.at(i, k, j)-ch3.at(i, k, jc))
+						c13.set(i, k, j, ch3.at(i, k, j)+ch3.at(i, k, jc))
+						c13.set(i, k, jc, ch3.at(i-1, k, jc)-ch3.at(i-1, k, j))
+					}
+				}
+			}
+		} else {
+			for j := 2; j <= ipph; j++ {
+				jc := ipp2 - j
+				for k := 1; k <= l1; k++ {
+					for i := 3; i <= ido; i += 2 {
+						c13.set(i-1, k, j, ch3.at(i-1, k, j)+ch3.at(i-1, k, jc))
+						c13.set(i-1, k, jc, ch3.at(i, k, j)-ch3.at(i, k, jc))
+						c13.set(i, k, j, ch3.at(i, k, j)+ch3.at(i, k, jc))
+						c13.set(i, k, jc, ch3.at(i-1, k, jc)-ch3.at(i-1, k, j))
+					}
+				}
+			}
+		}
+	}
+
+	for j := 2; j <= ipph; j++ {
+		jc := ipp2 - j
+		for k := 1; k <= l1; k++ {
+			c13.set(1, k, j, ch3.at(1, k, j)+ch3.at(1, k, jc))
+			c13.set(1, k, jc, ch3.at(1, k, jc)-ch3.at(1, k, j))
+		}
+	}
+	ar1 := 1.0
+	ai1 := 0.0
+	for l := 2; l <= ipph; l++ {
+		lc := ipp2 - l
+		ar1h := dcp*ar1 - dsp*ai1
+		ai1 = dcp*ai1 + dsp*ar1
+		ar1 = ar1h
+		for ik := 1; ik <= idl1; ik++ {
+			ch2m.set(ik, l, c2m.at(ik, 1)+ar1*c2m.at(ik, 2))
+			ch2m.set(ik, lc, ai1*c2m.at(ik, ip))
+		}
+		dc2 := ar1
+		ds2 := ai1
+		ar2 := ar1
+		ai2 := ai1
+		for j := 3; j <= ipph; j++ {
+			jc := ipp2 - j
+			ar2h := dc2*ar2 - ds2*ai2
+			ai2 = dc2*ai2 + ds2*ar2
+			ar2 = ar2h
+			for ik := 1; ik <= idl1; ik++ {
+				ch2m.set(ik, l, ch2m.at(ik, l)+ar2*c2m.at(ik, j))
```

Comment:
These two lines suggest that a `twoArray.add(i, j int, v float64` method would be a worthwhile optimisation.

Reply:
There are done.